text/0000-cargo-asymmetric-tokens.md

```diff
@@ -0,0 +1,218 @@
+- The biscuit language has some pretty complicated primitives, including regular expressions. Registries that require thorough correctness audits for all code related to Auth may find this prohibitively expensive.
+- The current biscuit specification (2.0) does not have a rich model of authentication. If you have a token that was authorized to do the action you are attempting to do then you must be someone who is allowed to do that action. Which has a lot of the same limitations of the existing secret token system as outlined in the motivation section of this RFC.
+- It is still possible to do scopes for tokens without using biscuits. A user ID can be created for each authorized role, and then the server can make sure that the used user ID is authorized to do the intended action.
+
+# Prior art
+[prior-art]: #prior-art
+
+NuGet has support for [author signing](https://github.com/NuGet/Home/wiki/Author-Package-Signing), which can be used to make sure that publishes only happen from somebody who has a private key. This system allows authenticity to be checked looking only at the crate that is downloaded. 
+However, in order to participate the author must have a "code signing certificate" from a "trusted root authority", making the barrier to participation to high for most users and certainly too high to be considered a norm of the community.
+
+Maven Central [requires](https://maven.apache.org/repository/guide-central-repository-upload.html#pgp-signature) all uploads to be [signed by PGP](https://central.sonatype.org/publish/requirements/gpg/) and that the keys are registered with a public key server. Following the UNIX philosophy, they leave the actual act of signing up to independent implementations of PGP.
+It takes a lot of documentation to explain how to hook up all of these different parts to work together correctly. Furthermore, no assurance is made that the GPG signature and the Maven Central token used for upload represent the same identity.
+
+The npm client can pass along a `otp` option on the command line to act as [proof of 2FA](https://docs.npmjs.com/configuring-two-factor-authentication#sending-a-one-time-password-from-the-command-line). This provides a lot of the "over the wire" benefits of this RFC for the npm registry, but cannot be used by a third party after the fact to verify the uploaded identity nor is it easy for alternative registries to set up.
+
+[TUF](https://theupdateframework.io/) exclusively deals with how a client downloading packages through a mirror can be assured they came from a non-compromised copy of the registry. Which is not the problem this RFC is addressing.
+
+# Unresolved questions
```

Comment:
How do we handle tokens passed on the command line via `--token`. Will we also support passing a private key via command line?

Reply:
It is possible to set it using the environment variable. But we have a shorthand for setting `--token` so maybe we should have a shorthand for `private-key` and `private-key-subject`. Do you think it's reasonable to consider this a part of the unresolved question about CLI design, or should we resolve it in the text of this RFC?

Reply:
I think we should make a decision about it in the RFC.

I'd rather avoid passing private keys on the command line since it increases the chances of them being exposed.

Reply:
The RFC no longer mentions passing keys on the command line.
The existing commands that take a `--token`, if it is set it will override any `private-key`/`private-key-subject` from the config.toml or environment variables. That's the only thing that maintains backwards compatibility.